include/seqan3/io/stream/iterator.hpp

```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+     * may use memcpy if applicable. Otherwise, a simple for loop iterates over the chunk.
+     *
+     * \attention You can only use the return value (end iterator) if your range type models
+     *            `std::ranges::forwarding_range`.
+     *
+     * Example:
+     *
+     * \include test/snippet/io/detail/iterator_write_range.cpp
+     */
+    template <std::ranges::forward_range range_type>
+    //!\cond
+        requires forwarding_range<range_type>
+    //!\endcond
+    auto write_range(range_type && rng)
+    {
+        using sen_t = std::ranges::sentinel_t<range_type>;
+        using it_t = std::ranges::iterator_t<range_type>;
+
+        it_t rng_it = std::ranges::begin(rng);
+        sen_t rng_end = std::ranges::end(rng);
+
+        size_t const buffer_space = stream_buf->epptr() - stream_buf->pptr();
+
+        if constexpr (std::ranges::sized_range<range_type>)
+        {
+            size_t const characters_to_write = std::min<size_t>(std::ranges::size(rng), buffer_space);
+            auto copy_res = std::ranges::copy_n(rng_it, characters_to_write, stream_buf->pptr());
+            rng_it = copy_res.in;
+            stream_buf->pbump(characters_to_write);
+        }
+        else
+        {
+            for (size_t i = 0; rng_it != rng_end && i < buffer_space; ++rng_it, ++i)
+            {
+                *stream_buf->pptr() = *rng_it;
+                stream_buf->pbump(1);
+            }
+        }
+
+        if (rng_it == rng_end) // no more characters to write
+            return rng_it;
+
+        // Push one more character and flush
+        if (stream_buf->overflow(*rng_it) == traits_t::eof())
+            throw std::ios_base::failure{"Cannot write to output stream (reached traits::eof() condition)."};
+
+        ++rng_it; // drop 1 character that has been written in overflow()
+
+        // Write remaining range recursively. For sized ranges, ensure that the subrange is also sized.
+        if constexpr (std::ranges::sized_range<range_type>)
+        {
+            // Note: Using ranges here because of bug https://github.com/ericniebler/range-v3/pull/1415 (fixed)
+            // TODO: change to std::ranges::subrange, when updating to range-v3 > 0.10.0
+            using subrange_t = ranges::subrange<it_t, sen_t, ranges::subrange_kind::sized>;
+            size_t const remaining_size = std::ranges::size(rng) - buffer_space - 1;
+            return write_range(subrange_t{rng_it, rng_end, remaining_size});
+        }
+        else
+        {
+            return write_range(std::ranges::subrange<it_t, sen_t>{rng_it, rng_end});
+        }
+    }
+
+    //!\cond
+    // overload for non-forwarding_range types that return void
+    template <std::ranges::forward_range range_type>
+    void write_range(range_type && rng)
+    {
+        write_range(rng); // return value is ignored because iterator would be dangling
```

Comment:
I added a test

Reply:
Sorry for the noise, but I can't find the test (edit: Found it). And I still think that this construct is not possible.

edit: My guess is an infinite loop :+1:

Reply:
Well 
1. since there is a test you can see that it works?
2. An lvalue is always a safe range so why should this ever cause an infinite loop as `rng` whithin the function scope is now an lvalue

Reply:
Boah this is not intuitive at all; can you add a comment that rng is a lvalue and hence the `safe range` overload is taken